### DISCUSSION TITLE
Possible COG like behavior in FGB?

### DISCUSSION THREAD

**Discussion by PostholerCom** [flatgeobuf/flatgeobuf#375] [Ideas]
FlatGeobuf has been a real game changer for me and I cannot say enough good things about it (and I do!). Not only is it great used for its intended purpose, it's great for building [metadata catalogs](https://www.postholer.com/fgbcog) of your raster/vector data files.

Working with huge FGB data sets and large bounding boxes are the only real difficulty I've come across. For instance, say your interactive web map is at zoom level 6 and you're trying to load thousands or millions of polygons. It's just too much data to download/display in the browser.

The method I've been using is creating 2 different FGB files. One is high resolution and the other is low. The low res might be displayed at zoom 1-10 and the high gets displayed at zoom 11-20. For point geom the low res will have a subset of points. A subset of polygons will have reduced vertices in the low res. This has been working very well. 

It's important to get FGB data straight from a web server or cloud storage *without* intermediate servers or services.

Not knowing enough about the underlying FGB spec, here's my question:

Could one or more additional rtree's be added to the FGB file that points to geom in the data that would simulate a low res chunk of that data? This would be similar to what COG does, but instead of creating a low res tile at some level, the additional rtree would reference its low res data from the existing data using another rtree. EDIT: even adding another data block with the additional rtree would work.

Ideally, this would allow you to stream a chunk of data, defined by bbox, at a specified level. 

Does this make any sense at all? If you made it this far, thanks for listening!

**Comment by michaelkirk** [comment]
Are you removing features in the low res version or are you keeping all the features and simplifying the geometries?

**Reply by PostholerCom** [reply]
Both. It depends on the size of the polygon. No sense having a building footprint at zoom 5, as it make no sense to have every vertex of Lake Tahoe at zoom 5.

**Reply by michaelkirk** [reply]
That makes sense, though I suppose the simplification logic might be somewhat application specific.

The short version is I can't think of a way to do what you're asking without breaking the format in a couple different ways. And those ways aren't much better than just having separate FGB's like you already are.

> Could one or more additional rtree's be added to the FGB file that points to geom in the data that would simulate a low res chunk of that data? 

I'm not sure what you mean by "simulate a low res chunk of that data". Currently the file only contains full resolution vector data. 

It'd be possible (by breaking the format) to add a separate index that points to a subset of these features, but they will all be full resolution features.

It'd be possible (by breaking the format) to write low resolution features in a separate feature buffer and then have the aforementioned low resolution index point to those low resolution features, but at that point, I'm not sure you've gained much vs. what you are currently doing: having a high resolution FGB and a low resolution FGB.

**Reply by michaelkirk** [reply]
I guess one thing that could be gained from storing the simplifications in the same file is that they could, in theory, reference the same non-geometry attributes ("properties"), so you wouldn't need to copy the non-geometry data for each resolution level, giving you some space savings over having separate FGBs.

Currently though, the geometry and non-geometry properties are stored adjacently, so that'd be (another) big breaking change from the current format.

**Reply by michaelkirk** [reply]
I wonder if you've looked at pmtiles?

https://github.com/protomaps/PMTiles

Sounds like it might be closer to what you're looking for.

**Reply by PostholerCom** [reply]
Hey Michael, I appreciate the responses! I realize what I was suggesting was a big deal, but I thought I'd throw it out there.

FGB made it possible to move away from any sort of tiling debacle, ie, PMTiles. Maintaining, updating tiles for zoom levels 1-18 with a national dataset is only for those with resources to deal with it, which is few of us. With PMTiles you can only get the vector data for a bbox of a tile, not the underlying geometry, without some server side implementation. Try clipping a raster in browser with vector data from a PMTIle, it's not pretty. With FGB it's a no-brainer. FGB also allows easy analysis of 2 or more vector data sets in browser. Again, with PMTiles it's a nightmare, if even possible without server side help.

With the availability of FGB, vector tiles make zero sense to me.

**Comment by bjornharrtell** [comment]
Hi @PostholerCom, when I designed FlatGeobuf I wanted to focus on simplicity and lossless 1-1 representation. You seem to desire some level of detail support which I see as out of scope. I suppose FlatGeobuf can replace vector tiles in some cases but it wasn't made as a replacement. The inherent tiling and generalization that is part of vector tiles make it suitable for rendering at scale, that was not a goal for me.

However, handling level of detail outside of the format like you do seems reasonable to me. Perhaps such use can be standardized and even be evolved into the format but it's not something I have plans to do at this point.

**Reply by PostholerCom** [reply]
Thanks for the feedback Bjorn. Again, just thought I'd see what everyone thought. 

Tile schemes like WMTS will create a tile if it doesn't exist, which is efficient but not portable due to server side processing.

Vector tiles you have to create every tile for every zoom level up front. This is woefully inefficient and a nightmare to maintain for *large* tile sets. This isn't an option for me. I would never commit resources to maintaining it.

With vector data I need to interact with, the high/low resolution FGB works well. If I don't need to interact with the low res polygon data, I just rasterize it and use COG. 

I can't think of a single valid reason to use vector tiles on *large* data sets, especially when you have FGB.

**Reply by mtravis** [reply]
@PostholerCom I'm intrigued to why you can't get along with vector tiles. How large are your datasets?

We use tippecanoe to create vector tiles and it find it works well for us in terms of only selecting certain zoom levels, but then we're only using for visual purposes as intended.

**Reply by bjornharrtell** [reply]
I'm not sure I agree with your statements @PostholerCom. It's possible to fx. setup pg_tileserv to do on demand tiles. I do not see FlatGeobuf as a direct competitor to vector tiles (also stated in as a FAQ long time ago here), to me these formats serve different purposes. That said, in some cases FlatGeobuf can work well for on demand fetch and render scenarios.

**Reply by PostholerCom** [reply]
@mtravis the largest data sets I use are CONUS FEMA vector flood zones and vector building footprint. 39GB and 36GB. [Give it a try](https://www.femafhz.com/map/27.937849/-82.461072/16/femafhz,footprints). Click on a building foot print.

Using the techniques above, fema data is rasterized and used for zoom levels 1-12, zoom 13+ FGB vector data is displayed. The FGB building footprints are only shown at zoom 15+. So, 1 COG file and 2 FGB files.

The important feature is, I can click on a building footprint, get its area and the flood zone(s) it occupies. No server side processing, only the FGB files sitting on a web server. *All done in the browser.*

FEMA does regular, daily updates. Using ogr2ogr I upsert any new data to a .gpkg, gdal_rasterize for new raster, then ogr2ogr for new FGB. This massive FGB fema dataset gets updated nightly. 

I couldn't imagine doing this with vector tiles. Tippacanoe, probably python, who knows what else, creating vector tiles for zoom 13-18 for all of CONUS, nightly. Even worse, I can't interact with underlying vector data (flood polygons) in browser without some server side assistance.

Spatial visualization in a cloud native manner, meaning no server/services processing, is the holy grail. FGB and COG make this possible. Visualization and light analysis in your browser is utterly simple. I'm shocked more people haven't run to this with open arms.

**Reply by PostholerCom** [reply]
@bjornharrtell pg_tileserv is server side processing. It negates the whole concept of cloud native data formats. For *visualization* of cloud native data/formats it defeats the purpose.

Again, vector tiles are fine for small data sets and/or small range of zoom levels. Vector tiles are a nightmare for large areas and a wide range of zoom levels. If you need to do light analysis on the underlying vector data, it's impossible without server side.

**Reply by bjornharrtell** [reply]
Happy you are happy 🙂 I do have some similar use in production here and there and it can definitely work well, just don't want to claim it is a silver bullet.

**Reply by PostholerCom** [reply]
@bjornharrtell I don't think you realize just how awesome your FGB format is and what it's capable of. I'm sure glad you put it together!

**Reply by mtravis** [reply]
> @mtravis the largest data sets I use are CONUS FEMA vector flood zones and vector building footprint. 39GB and 36GB. [Give it a try](https://www.femafhz.com/map/27.937849/-82.461072/16/femafhz,footprints). Click on a building foot print.
> 
> Using the techniques above, fema data is rasterized and used for zoom levels 1-12, zoom 13+ FGB vector data is displayed. The FGB building footprints are only shown at zoom 15+. So, 1 COG file and 2 FGB files.
> 
> The important feature is, I can click on a building footprint, get its area and the flood zone(s) it occupies. No server side processing, only the FGB files sitting on a web server. _All done in the browser._
> 
> FEMA does regular, daily updates. Using ogr2ogr I upsert any new data to a .gpkg, gdal_rasterize for new raster, then ogr2ogr for new FGB. This massive FGB fema dataset gets updated nightly.
> 
> I couldn't imagine doing this with vector tiles. Tippacanoe, probably python, who knows what else, creating vector tiles for zoom 13-18 for all of CONUS, nightly. Even worse, I can't interact with underlying vector data (flood polygons) in browser without some server side assistance.
> 
> Spatial visualization in a cloud native manner, meaning no server/services processing, is the holy grail. FGB and COG make this possible. Visualization and light analysis in your browser is utterly simple. I'm shocked more people haven't run to this with open arms.

Nice. That's a cool use of FGBs!
